# Problem Set 1, due January 13th at 12:00pm 
(in other words, it is due at noon, or one minute after 11:59am).

###Before You Start
Make sure to at least take a basic tutorial in the IPython notebook, otherwise you'll be totally lost.  For this problem set, you should download INFX574-PS1.ipynb and the flights.zip dataset from Canvas. Create a local copy of the notebook and rename it LASTNAME_FIRSTNAME-PS1.ipynb. Then edit your renamed file directly in your browser by typing:
```
ipython notebook <name_of_downloaded_file>
```

You should also make sure the following libraries load correctly (click on the box below and hit Ctrl-Enter)

In [ ]:
#Name: Shuyang Wu
#Date: 1/13/2017
#Section: C

In [2]:
#IPython is what you are using now to run the notebook
import IPython
print ("IPython version:      %6.6s (need at least 1.0)" % IPython.__version__)

# Numpy is a library for working with Arrays
import numpy as np
print ("Numpy version:        %6.6s (need at least 1.7.1)" % np.__version__)

# SciPy implements many different numerical algorithms
import scipy as sp
print ("SciPy version:        %6.6s (need at least 0.12.0)" % sp.__version__)

# Pandas makes working with data tables easier
import pandas as pd
print ("Pandas version:       %6.6s (need at least 0.11.0)" % pd.__version__)

# Module for plotting
import matplotlib
print ("Mapltolib version:    %6.6s (need at least 1.2.1)" % matplotlib.__version__)

# SciKit Learn implements several Machine Learning algorithms
import sklearn
print ("Scikit-Learn version: %6.6s (need at least 0.13.1)" % sklearn.__version__)

IPython version:       4.1.2 (need at least 1.0)
Numpy version:        1.10.4 (need at least 1.7.1)
SciPy version:        0.17.0 (need at least 0.12.0)
Pandas version:       0.18.0 (need at least 0.11.0)
Mapltolib version:     1.5.1 (need at least 1.2.1)
Scikit-Learn version: 0.17.1 (need at least 0.13.1)


##About the Problem Set: 
This is the same problem set used by Emma Spiro in INFX573. The only difference is that instead of doing the problem set in R, you will use Python and the IPython notebook.

##Instructions: 
In this problem set you will perform a basic exploratory analysis on an example dataset, bringing to bear all of your new skills in data manipulation and visualization. You will be required to submit well commented python code, documenting all code used in this problem set, along with a write up answering all questions below. Use figures as appropriate to support your answers, and when required by the problem. 
This data set uses the NYCFlights13 dataset. You can download the dataset from canvas.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [119]:
flights_df= pd.read_csv('~/Downloads/flights.csv')

In [5]:
print (flights_df.shape)
print (flights_df.columns)
print (flights_df.dtypes)

(336776, 17)
Index(['Unnamed: 0', 'year', 'month', 'day', 'dep_time', 'dep_delay',
       'arr_time', 'arr_delay', 'carrier', 'tailnum', 'flight', 'origin',
       'dest', 'air_time', 'distance', 'hour', 'minute'],
      dtype='object')
Unnamed: 0      int64
year            int64
month           int64
day             int64
dep_time      float64
dep_delay     float64
arr_time      float64
arr_delay     float64
carrier        object
tailnum        object
flight          int64
origin         object
dest           object
air_time      float64
distance        int64
hour          float64
minute        float64
dtype: object


In [6]:
flights_df.dest.unique()
flights_df.head(10)

Unnamed: 0  year  month  day  dep_time  dep_delay  arr_time  arr_delay  \
0           1  2013      1    1     517.0        2.0     830.0       11.0   
1           2  2013      1    1     533.0        4.0     850.0       20.0   
2           3  2013      1    1     542.0        2.0     923.0       33.0   
3           4  2013      1    1     544.0       -1.0    1004.0      -18.0   
4           5  2013      1    1     554.0       -6.0     812.0      -25.0   
5           6  2013      1    1     554.0       -4.0     740.0       12.0   
6           7  2013      1    1     555.0       -5.0     913.0       19.0   
7           8  2013      1    1     557.0       -3.0     709.0      -14.0   
8           9  2013      1    1     557.0       -3.0     838.0       -8.0   
9          10  2013      1    1     558.0       -2.0     753.0        8.0   

  carrier tailnum  flight origin dest  air_time  distance  hour  minute  
0      UA  N14228    1545    EWR  IAH     227.0      1400   5.0    17.0  
1      UA  N24211    1714    LGA  IAH     227.0      1416   5.0    33.0  
2      AA  N619AA    1141    JFK  MIA     160.0      1089   5.0    42.0  
3      B6  N804JB     725    JFK  BQN     183.0      1576   5.0    44.0  
4      DL  N668DN     461    LGA  ATL     116.0       762   5.0    54.0  
5      UA  N39463    1696    EWR  ORD     150.0       719   5.0    54.0  
6      B6  N516JB     507    EWR  FLL     158.0      1065   5.0    55.0  
7      EV  N829AS    5708    LGA  IAD      53.0       229   5.0    57.0  
8      B6  N593JB      79    JFK  MCO     140.0       944   5.0    57.0  
9      AA  N3ALAA     301    LGA  ORD     138.0       733   5.0    58.0

##Some Tips

* This assignment involves extensive Data frame splitting and aggregation. You should look into the details of the methods groupby, transform, sum, count, mean etc
* Many of the tasks in the assignment can be done either through the Pandas Data Frame or by converting the data frames to Series. Many of the methods in the numpy are applicable to Series only. When stuck, try to explore the type of object (Pandas Data Frame or Numpy Series) you are dealing with.

##Question 1
Let’s explore flights from NYC to Seattle. Use the flights dataset to answer the following questions.

(a) How many flights were there from NYC airports to Seattle in 2013?

In [11]:
# Your code here
sea = flights_df[flights_df["dest"] == "SEA"]
print("Number of flights from NYC to SEA: ", sea.size)

Number of flights from NYC to SEA:  66691


 -- Write your answer in English here -- 
 
 There were 66691 flights from NYC airports to Seattle in 2013.

(b) How many airlines fly from NYC to Seattle?

In [15]:
# Your code here
sea.carrier.nunique()

5

 -- Write your answer in English here -- 
 
 5 different airlines fly from NYC to Seattle

(c) How many unique air planes fly from NYC to Seattle?

In [23]:
# Your code here
sea.groupby("carrier").flight.nunique().sum()

166

 -- Write your answer in English here -- 
 
 There were 166 unique airplanes flew from NYC to Seattle.

(d) What is the average arrival delay for flights from NC to Seattle?

In [29]:
# Your code here
sea.arr_delay.mean()

-1.0990990990990992

 -- Write your answer in English here -- 
 
 The average arrival delay for flights from NYC to Seattle was about -1.1 minutes.

(e) What proportion of flights to Seattle come from each NYC airport?

In [37]:
# Your code here
sea.groupby("origin").size()/(sea.origin.size)

origin
EWR    0.466735
JFK    0.533265
dtype: float64

 -- Write your answer in English here -- 
 
 46.7% of flights came from EWR airport, 53.3% of flights came from JFK airport.

## Question 2 (two ways of answering per question)
Flights are often delayed. Consider the following questions exploring delay patterns.

(a) Which date has the largest average departure delay? Which date has the largest average arrival delay?

In [67]:
#First approach
ave_dep = flights_df.groupby(["year","month","day"]).dep_delay.mean() #groupby date, get the mean delay for each date
print(ave_dep.idxmax()) #get the index of the max value
ave_arr = flights_df.groupby(["year","month","day"]).arr_delay.mean()
print(ave_arr.idxmax())

#Second approach
ave_dep1 = flights_df.groupby(["year","month","day"]).dep_delay.mean()
print(ave_dep1.sort_values().index[-1]) #sort in ascending order, then get the index of the last element
ave_arr1 = flights_df.groupby(["year","month","day"]).arr_delay.mean()
print(ave_arr1.sort_values().index[-1])

(2013, 3, 8)
(2013, 3, 8)
(2013, 3, 8)
(2013, 3, 8)


 -- Write your answer in English here -- 
 
 2013/3/8 had both the largest average departure delay and arrival delay.

(b) What was the worst day to fly out of NYC in 2013 if you dislike delayed flights?


In [ ]:
# Same as question 2(a)
#First approach
ave_dep = flights_df.groupby(["year","month","day"]).dep_delay.mean() #groupby date, get the mean delay for each date
print(ave_dep.idxmax()) #get the index of the max value
ave_arr = flights_df.groupby(["year","month","day"]).arr_delay.mean()
print(ave_arr.idxmax())

#Second approach
ave_dep1 = flights_df.groupby(["year","month","day"]).dep_delay.mean()
print(ave_dep1.sort_values().index[-1]) #sort in ascending order, then get the index of the last element
ave_arr1 = flights_df.groupby(["year","month","day"]).arr_delay.mean()
print(ave_arr1.sort_values().index[-1])

 -- Write your answer in English here -- 
 2013/3/8 was the worst day to fly out of NYC if you dislike delayed flights.

(c) Are there any seasonal patterns in departure delays for flights from NYC?

In [145]:
# Your code here
#First Approach
#Assign seasons to months
flights_df['season'] = pd.Series(np.random.randn(len(flights_df["month"])), index=flights_df.index)
flights_df.loc[(flights_df["month"] == 12)| (flights_df["month"] == 1)|(flights_df["month"] == 2), ["season"]] = "Winter"
flights_df.loc[(flights_df["month"] == 3) | (flights_df["month"] == 4)|(flights_df["month"] == 5), ["season"]] = "Spring"
flights_df.loc[(flights_df["month"] == 6) | (flights_df["month"] == 7) |(flights_df["month"] == 8), ["season"]] = "Summer"
flights_df.loc[(flights_df["month"] == 9) | (flights_df["month"] == 10) | (flights_df["month"] == 11), ["season"]] = "Fall"
#groupby seasons
print(flights_df.groupby("season").dep_delay.mean())

#Second Approach
#groupby months
mon_delay = flights_df.groupby("month").dep_delay.mean()
index = ['winter', 'winter','spring', 'spring', 'spring', 'summer', 'summer', 'summer', 'fall', 'fall','fall', 'winter']
#reindexing months into seasons
mon_delay.index = (index)
print(mon_delay)
#show pattern
mon_delay.plot()
plt.show()

season
Fall       6.136711
Spring    13.380705
Summer    18.334437
Winter    12.569983
Name: dep_delay, dtype: float64
winter    10.036665
winter    10.816843
spring    13.227076
spring    13.938038
spring    12.986859
summer    20.846332
summer    21.727787
summer    12.611040
fall       6.722476
fall       6.243988
fall       5.435362
winter    16.576688
Name: dep_delay, dtype: float64


 -- Write your answer in English here -- 
 
 Summer tends to have the longest delay time which was 18 minutes on average.
 Fall tends to have the shortest delay time which was 6 minutes on average.

(d) On average, how do departure delays vary over the course of a day?

In [154]:
# Your code here
#First Approach, use line plot to show pattern
timevsdelay = flights_df.groupby("dep_time").dep_delay.mean()
np.corrcoef(timevsdelay.index,timevsdelay.values) #correlation coefficient between time of the day and delay
timevsdelay.plot(kind = "line")
plt.show()

#Second Approach
#categorize time of a day
flights_df['m/n'] = pd.Series(np.nan, index=flights_df.index)
flights_df.loc[(flights_df["hour"] >= 6.0)&(flights_df["hour"] <= 12.0), ["m/n"]] = "Morning"
flights_df.loc[(flights_df["hour"] >= 13.0)&(flights_df["hour"] <= 18.0), ["m/n"]] = "Afternoon"
flights_df.loc[(flights_df["hour"] >= 19.0)&(flights_df["hour"] <= 23.0), ["m/n"]] = "Evening"
flights_df.loc[(flights_df["hour"] >= 24.0)|(flights_df["hour"] <= 5.0), ["m/n"]] = "Night"
flights_df.groupby("m/n").dep_delay.mean()

m/n
Afternoon    13.035694
Evening      35.786907
Morning       2.788139
Night        16.764357
Name: dep_delay, dtype: float64

 -- Write your answer in English here -- 
 
Departure delays on average tend to be much higher during the evenings (7pm - 11pm, highest) and nights, with a sudden decrease before 5 am. Mornings have very small mean delay time, 2.7 minutes.


## Question 3
    Which flight departing NYC in 2013 flew the fastest?

In [88]:
# Your code here
print(flights_df["carrier"][flights_df.air_time.min()])
print(flights_df["flight"][flights_df.air_time.min()])

DL
1919


 -- Write your answer in English here -- 
 
 DL 1919 flew the fastest in 2013.

## Question 4
Which flights (i.e. carrier + flight + dest) happen every day? Where do they fly to?

In [168]:
# Your code here
#get total number of days 
flights_df["date"] = flights_df["month"].map(str) + flights_df["day"].map(str)
ndays = flights_df.date.nunique()
print(ndays)
cfd = flights_df.groupby(["carrier","flight","dest"]).date.nunique() #get number of days each flight flew
cfd[cfd == ndays] 

347


carrier  flight  dest
AA       59      SFO     347
         119     LAX     347
         181     LAX     347
         1357    SJU     347
         1611    MIA     347
B6       219     CLT     347
         359     BUR     347
         371     FLL     347
         431     SRQ     347
         703     SJU     347
         1783    MCO     347
DL       2159    MCO     347
         2391    TPA     347
EV       5712    IAD     347
UA       15      HNL     347
VX       251     LAS     347
         407     LAX     347
         411     LAX     347
         413     LAX     347
Name: date, dtype: int64

 -- Write your answer in English here -- 
 
 The flight numbers and destinations of everyday flights are:
 
 AA59 to SFO, AA119 to LAX, AA181 to LAX, AA1357 to SJU, AA1611 to MIA;
 
 B6219 to CLT, B6359 to BUR, B6371 to FLL, B6431 to SRQ, B6703 to SJU, B61783 to MCO;
 
 DL2159 to MCO, DL2391 to TPA;
 
 EV5712 to IAD;
 
 UA15 to HNL;
 
 VX251 to LAS, VX407 to LAX, VX411 to LAX, VX413 to LAX;

## Question 5
Develop one research question you can address using the nycflights2013 dataset. Provide two visualizations to support your exploration of this question. Discuss what you find.


In [192]:
# Your code here
#research questions: Is there a pattern for flights dispatching in regards to the days/months?
#plot the number of flights leaving NYC everyday
daily = flights_df.groupby(["year","month","day"]).flight.count()
daily.plot(kind = "line")
plt.show()
#plot the number of flights leaving NYC every month
monthly = flights_df.groupby(["year","month"]).flight.count()
monthly.plot(kind = "bar")
plt.ylim(ymin=24000)
plt.ylim(ymax=30000)
plt.show()
print(monthly.sort_values())
seasonly = flights_df.groupby("season").flight.count()
seasonly.plot(kind = "bar")
plt.ylim(ymin=75000)
plt.ylim(ymax=90000)
plt.show()
print(seasonly.sort_values())

year  month
2013  2        24951
      1        27004
      11       27268
      9        27574
      12       28135
      6        28243
      4        28330
      5        28796
      3        28834
      10       28889
      8        29327
      7        29425
Name: flight, dtype: int64
season
Winter    80090
Fall      83731
Spring    85960
Summer    86995
Name: flight, dtype: int64


 -- Enter your discussion here --
 
There seem to be quite large oscillations in terms of number of flights leaving NYC vary from day to day. On average, February had the smallest number of flights, whereas July had the largest. More generally speaking, there were fewer flights in winter and more flights in summer.

## Question 6
What weather conditions are associated with flight delays leaving NYC? Use graphics to explore.

In [ ]:
# Your code here
#look at the dataset
weather= pd.read_csv('~/Downloads/weather.csv')
print (weather.shape)
print (weather.columns)
print (weather.dtypes)
#print (weather.head(10))

#Combine two datasets into df
flights_d = flights_df[flights_df.dep_delay >= 30] #keep only the flights whose delay time is over 30 mins
dd = pd.concat([flights_df['month'],flights_df['day'], flights_df['dep_delay']], axis=1)
df = pd.merge(dd, weather, on=['month','day'], how='outer')
df = df.drop(df.columns[[0, 1, 3,4,5,6]], axis=1) #keep only the weather-related columns
#correlation coefficient between departure delay and all weather related information
print(df.corr()['dep_delay']) 

#scatterplot of the association between departure delay and humidity
m, b = np.polyfit(df.humid, df.dep_delay, 1)
plt.plot(df.humid, df.dep_delay, '.')
plt.plot(df.humid, m*df.humid + b, '-')
plt.show()

(8719, 15)
Index(['Unnamed: 0', 'origin', 'year', 'month', 'day', 'hour', 'temp', 'dewp',
       'humid', 'wind_dir', 'wind_speed', 'wind_gust', 'precip', 'pressure',
       'visib'],
      dtype='object')
Unnamed: 0      int64
origin         object
year            int64
month         float64
day           float64
hour          float64
temp          float64
dewp          float64
humid         float64
wind_dir      float64
wind_speed    float64
wind_gust     float64
precip        float64
pressure      float64
visib         float64
dtype: object
dep_delay     1.000000
temp          0.058980
dewp          0.098020
humid         0.106202
wind_dir     -0.041077
wind_speed   -0.002850
wind_gust    -0.002850
precip        0.044321
pressure     -0.071745
visib        -0.089652
Name: dep_delay, dtype: float64


-- Enter your interpretation here --

There seem to be no strong association between weather condition and delayed departure time. The strongest correlation amongst all is between delayed departure time and humidity, with a correlation coefficient of 0.106. Scatterplot also show little evidence of significant association.